tests/migration/model.go

```diff
@@ -0,0 +1,137 @@
+package migration
+
+import (
+	"bytes"
+	"testing"
+)
+
+// ItemGroupModel describes migration model of some storage item group of the
+// fixed Neo contract.
+type ItemGroupModel struct {
+	keyPattern binPattern
+	modifier   func(prev kv) kv
+}
+
+// UpdateModel describes storage migration model of some Neo contract. Zero update model
+// describes empty post-update contract storage.
+type UpdateModel struct {
```

Comment:
This model describes change set in terms of KV pairs, which is fine when:
 * we delete some key or a set of keys with some prefix and want to ensure it's gone
 * we specifically add some key or a set of keys and want to ensure they're present
 * we want to make sure no excessive elements are added

But it's suboptimal when we're dealing with changes. Consider our netmap 0.16.0 migration problem (df2bad0f2fcddc1d405be3fa8003eec3c8d1d1b1), if we have this omission in the contract we're quite likely to have the same omission in the test, but calling contract APIs could quickly reveal the problem, so I'd say we should describe these in terms of API calls (test invocations) that should have the same (or intentionally different if needed) result before the migration and after the migration.

Reply:
IMO these are two different tests:
1. migration of storage items from one format to another
2. ability of new contract executable to correctly decode new format

2nd won't help us if there is no corresponding values in the contract at all. But I agree that such API extra check will help us to feel more comfortable.